File: Flood fill Algorithm - GFG/flood-fill-algorithm.java
//{ Driver Code Starts
import java.util.*;
import java.lang.*;
import java.io.*;
class GFG
{
    public static void main(String[] args) throws IOException
    {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int T = Integer.parseInt(br.readLine().trim());
        while(T-->0)
        {
            String[] S1 = br.readLine().trim().split(" ");
            int n = Integer.parseInt(S1[0]);
            int m = Integer.parseInt(S1[1]);
            int[][] image =  new int[n][m];
            for(int i = 0; i < n; i++){
                String[] S2 = br.readLine().trim().split(" ");
                for(int j = 0; j < m; j++)
                    image[i][j] = Integer.parseInt(S2[j]);
            }
            String[] S3 = br.readLine().trim().split(" ");
            int sr = Integer.parseInt(S3[0]);
            int sc = Integer.parseInt(S3[1]);
            int newColor = Integer.parseInt(S3[2]);
            Solution obj = new Solution();
            int[][] ans = obj.floodFill(image, sr, sc, newColor);
            for(int i = 0; i < ans.length; i++){
                for(int j = 0; j < ans[i].length; j++)
                    System.out.print(ans[i][j] + " ");
                System.out.println();
            }
        }
    }
}

// } Driver Code Ends


class Solution
{
    private void dfs(int row,int col,int ans[][],int image[][],int newColor,
                     int intialColor,int delRow[],int delCol[]){
        ans[row][col] = newColor;
        int n = image.length;
        int m = image[0].length;
        for(int i =0;i<4;i++){
            int nrow = row + delRow[i];
            int ncol = col + delCol[i];
            if(nrow>=0 && nrow < n && ncol >= 0 && ncol < m && image[nrow][ncol] == intialColor
            && ans[nrow][ncol] != newColor){
                dfs(nrow,ncol,ans,image,newColor,intialColor,delRow,delCol);
            }
        }
        
    }
    public int[][] floodFill(int[][] image, int sr, int sc, int newColor)
    {
       int intialColor = image[sr][sc];
       int ans[][] = image;
      int delRow[] = {-1,0,+1,0};
      int delCol[] = {0,+1,+0,-1};
        dfs(sr,sc,ans,image,newColor,intialColor,delRow,delCol);
        return ans;
    }
}

File: Flood fill Algorithm - GFG/flood-fill-algorithm.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends
class Solution {
private:
    void dfs(int row,int col, int initial_color,int newColor,vector<vector<int>>& ans,
        vector<vector<int>>& image,int delRow[],int delCol[]) {
        ans[row][col] = newColor;
        int n = image.size();
        int m = image[0].size();
        for(int i = 0; i < 4; i++) {
            int nrow = row + delRow[i];
            int ncol = col + delCol[i];
            if(nrow >= 0 && nrow < n && ncol >= 0 && ncol < m && 
            image[nrow][ncol] == initial_color && ans[nrow][ncol] != newColor) {
                dfs(nrow,ncol,initial_color,newColor,ans,image,delRow,delCol);
            }
        }
    }
public:
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, 
        int newColor) {
        int initial_color = image[sr][sc];
        vector<vector<int>> ans = image;
        int delRow[] = {-1,0,1,0};
        int delCol[] = {0,1,0,-1};
        dfs(sr,sc,initial_color,newColor,ans,image,delRow,delCol);
        return ans;
    }
};

//{ Driver Code Starts.
int main(){
	int tc;
	cin >> tc;
	while(tc--){
		int n, m;
		cin >> n >> m;
		vector<vector<int>>image(n, vector<int>(m,0));
		for(int i = 0; i < n; i++){
			for(int j = 0; j < m; j++)
				cin >> image[i][j];
		}
		int sr, sc, newColor;
		cin >> sr >> sc >> newColor;
		Solution obj;
		vector<vector<int>> ans = obj.floodFill(image, sr, sc, newColor);
		for(auto i: ans){
			for(auto j: i)
				cout << j << " ";
			cout << "\n";
		}
	}
	return 0;
}
// } Driver Code Ends